### PR TITLE
Add support for fargate launch_type in ecs_service module and ecs_taskdefinition module

### LINKED ISSUE #35607: Add launch_type parameter for ecs_service module
##### ISSUE TYPE

 - Feature Idea

##### COMPONENT NAME

ecs_service

##### ANSIBLE VERSION

```
$ ansible --version
ansible 2.4.2.0
  config file = None
  configured module search path = [u'/Users/ramonamis/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/ramonamis/.virtualenvs/devops/lib/python2.7/site-packages/ansible
  executable location = /Users/ramonamis/.virtualenvs/devops/bin/ansible
  python version = 2.7.10 (default, Jul 15 2017, 17:16:57) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.31)]
```

##### OS / ENVIRONMENT

N/A
##### SUMMARY

By default, the ecs_service module is launching services with launch_type EC2. At the moment it is impossible to use launch_type: "FARGATE" in order to migrate to fargate.

For more details you can check: https://docs.aws.amazon.com/cli/latest/reference/ecs/create-service.html

##### STEPS TO REPRODUCE

It could look something like this: 
```yaml
          - name: deploy service
            ecs_service:
              region: "{{ aws_region }}"
              name: "{{ ecs_task_family }}"
              load_balancers: "{{ ecs_task_load_balancers }}"
              launch_type: "FARGATE"
              state: present
              task_definition: "{{ ecs_task_family  }}"
              cluster: "{{ ecs_cluster_name }}"
              desired_count: "{{ ecs_task_desired_count }}"
```

##### EXPECTED RESULTS

Be able to deploy ecs services to fargate
##### ACTUAL RESULTS

I am not able to deploy ecs services with fargate

```
fatal: [localhost]: FAILED! => {"attempts": 10, "changed": false, "msg": "Unsupported parameters for (ecs_service) module: launch_type Supported parameters include: aws_access_key,aws_secret_key,client_token,cluster,delay,deployment_configuration,desired_count,ec2_url,load_balancers,name,placement_constraints,placement_strategy,profile,region,repeat,role,security_token,state,task_definition,validate_certs"}
```

### PR DESCRIPTION
**Pull request by mjmayer**
##### SUMMARY
Add support for fargate launch_type in ecs_service module and ecs_taskdefinition module. This fixes #35607 . There are now two types of ECS clusters in AWS, EC2 and Fargate ,this adds support for Fargate.

Added some additional error checking to make sure when the playbook is executed it identifies required values such as memory, cpu, and network_type, and network_configuration before reaching out to the AWS API.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
ecs_service, ecs_taskdefinition

##### ANSIBLE VERSION
```
ansible 2.6.0dev0 (devel eb9fe31e51) last updated 2018/05/25 14:26:56 (GMT -700)
  config file = None
  configured module search path = [u'/home/mjmayer/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/mjmayer/ansible_mjmayer/lib/ansible
  executable location = /home/mjmayer/ansible_mjmayer/bin/ansible
  python version = 2.7.5 (default, Apr 11 2018, 17:41:36) [GCC 4.8.5 20150623 (Red Hat 4.8.5-28.0.1)]

```

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py]: These could be required_if in AnsibleModule

```
required_if(['launch_type', 'FARGATE', ['cpu', 'memory']])
```

You'll need your existing network_mode check

File: lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py]: This doesn't seem to have any checks for whether boto3 supports `requiresCompatibilities`, which seems to only come with version 1.8.4 and beyond of botocore.

You'll need a check similar to https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/ecs_service.py#L443-L445

`requiresCompatibilities` should only be added to the parameter list if it's actually required.

File: lib/ansible/modules/cloud/amazon/ecs_service.py
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ecs_service.py]: This doesn't seem to have any checks for whether boto3 supports `launchType`, which seems to only come with version 1.8.4 and beyond of botocore.

You'll need a check similar to https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/ecs_service.py#L443-L445

`launchType` should only be passed here if it's actually required.

**Review by willthames — CHANGES_REQUESTED** [review]

**Comment by mjmayer** [comment]
@willthames I've made the requested changes and update this PR.

File: lib/ansible/modules/cloud/amazon/ecs_service.py
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ecs_service.py]: This should read launchType, not placementStrategy

File: lib/ansible/modules/cloud/amazon/ecs_service.py
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ecs_service.py]: don't set a default here, or this won't work with `< 1.8.4`

File: lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py]: Don't set a default for `launch_type`

File: lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py]: This can just be `cpu=dict()` as both `required=False` and `type='str'` are the default.

**Review by willthames — CHANGES_REQUESTED** [review]
The main blocking issue now is just the default for `launch_type`.

However, that does identify for me that there are no tests associated with this change. 

The test for < 1.8.4 can be added to test/integration/targets/ecs_cluster/playbooks/network_fail.yml

The test for launch_type=fargate should be added to test/integration/targets/ecs_cluster/playbooks/full_test.yml

Let me know if you have any problems with the tests, and I'll see if I can help you resolve them.

**Comment by mjmayer** [comment]
@willthames I have the ansible-test environment setup according to your [directions](http://willthames.github.io/2017/07/17/so-you-want-to-test-aws-modules-for-ansible.html). I am running into failures of the tests for a previous PR #34381 
```
TASK [create ecs_service using awsvpc network_configuration] ******************************************************************************************************************************************************************************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: botocore.errorfactory.InvalidParameterException: An error occurred (InvalidParameterException) when calling the CreateService operation: Network Configuration is not valid for the given networkMode of this task definition.
fatal: [localhost]: FAILED! => {"changed": false, "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_vurQzX/ansible_module_ecs_service.py\", line 584, in <module>\n    main()\n  File \"/tmp/ansible_vurQzX/ansible_module_ecs_service.py\", line 532, in main\n    module.params['launch_type'])\n  File \"/tmp/ansible_vurQzX/ansible_module_ecs_service.py\", line 385, in create_service\n    response = self.ecs.create_service(**params)\n  File \"/home/vagrant/.local/lib/python2.7/site-packages/botocore/client.py\", line 314, in _api_call\n    return self._make_api_call(operation_name, kwargs)\n  File \"/home/vagrant/.local/lib/python2.7/site-packages/botocore/client.py\", line 612, in _make_api_call\n    raise error_class(parsed_response, operation_name)\nbotocore.errorfactory.InvalidParameterException: An error occurred (InvalidParameterException) when calling the CreateService operation: Network Configuration is not valid for the given networkMode of this task definition.\n", "module_stdout": "", "msg": "MODULE FAILURE", "rc": 1}
...ignoring

TASK [check that graceful failure message is returned from ecs_service] *******************************************************************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {
    "assertion": "ecs_service_creation_vpc.msg == \"botocore needs to be version 1.7.44 or higher to use network configuration\"",
    "changed": false,
    "evaluated_to": false
}
```
It seems like the module.fail_json() is not actually stopping the execution of the module.
https://github.com/ansible/ansible/blob/f87ca27aa3dc5a017374f2204ede1568dbcb4515/lib/ansible/modules/cloud/amazon/ecs_service.py#L443-L448

**Comment by willthames** [comment]
@mjmayer - thanks for addressing my feedback. I'll try and run these tests independently shortly.

I assume [the April 26 blog post](https://aws.amazon.com/about-aws/whats-new/2018/04/aws-fargate-now-available-in-ohio--oregon--and-ireland-regions/) is the latest fargate availability statement - what happens if the module is run outside of those regions?

**Comment by willthames** [comment]
Running against ap-southeast-2:

```
TASK [ecs_cluster : create fargate ECS service with network config] ************
task path: /root/ansible/test/integration/targets/ecs_cluster/playbooks/roles/ecs_cluster/tasks/main.yml:607
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: root
<127.0.0.1> EXEC /bin/sh -c 'echo ~root && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp/ansible-tmp-1528107543.5715554-127374206489247 `" && echo ansible-tmp-1528107543.5715554-127374206489247="` echo /root/.ansible/tmp/ansible-tmp-1528107543.5715554-127374206489247 `" ) && sleep 0'
Using module file /root/ansible/lib/ansible/modules/cloud/amazon/ecs_service.py
<127.0.0.1> PUT /root/.ansible/tmp/ansible-local-565dw2ul2qv/tmpchjwnni8 TO /root/.ansible/tmp/ansible-tmp-1528107543.5715554-127374206489247/ecs_service.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1528107543.5715554-127374206489247/ /root/.ansible/tmp/ansible-tmp-1528107543.5715554-127374206489247/ecs_service.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/tmp/ansible-test-coverage-5i8bs9_8/coverage/injector.py /root/.ansible/tmp/ansible-tmp-1528107543.5715554-127374206489247/ecs_service.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1528107543.5715554-127374206489247/ > /dev/null 2>&1 && sleep 0'
The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_xpehgsgp/ansible_module_ecs_service.py", line 583, in <module>
    main()
  File "/tmp/ansible_xpehgsgp/ansible_module_ecs_service.py", line 531, in main
    module.params['launch_type'])
  File "/tmp/ansible_xpehgsgp/ansible_module_ecs_service.py", line 384, in create_service
    response = self.ecs.create_service(**params)
  File "/usr/lib/python3.6/site-packages/botocore/client.py", line 314, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/usr/lib/python3.6/site-packages/botocore/client.py", line 612, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.errorfactory.UnsupportedFeatureException: An error occurred (UnsupportedFeatureException) when calling the CreateService operation: FARGATE is not supported in this region.

fatal: [localhost]: FAILED! => {
    "changed": false,
    "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_xpehgsgp/ansible_module_ecs_service.py\", line 583, in <module>\n    main()\n  File \"/tmp/ansible_xpehgsgp/ansible_module_ecs_service.py\", line 531, in main\n    module.params['launch_type'])\n  File \"/tmp/ansible_xpehgsgp/ansible_module_ecs_service.py\", line 384, in create_service\n    response = self.ecs.create_service(**params)\n  File \"/usr/lib/python3.6/site-packages/botocore/client.py\", line 314, in _api_call\n    return self._make_api_call(operation_name, kwargs)\n  File \"/usr/lib/python3.6/site-packages/botocore/client.py\", line 612, in _make_api_call\n    raise error_class(parsed_response, operation_name)\nbotocore.errorfactory.UnsupportedFeatureException: An error occurred (UnsupportedFeatureException) when calling the CreateService operation: FARGATE is not supported in this region.\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE",
    "rc": 1
}
```

The code just needs to handle the UnsupportedFeatureException in a friendly way. Testing this in a future-proof fashion might be tricky (for now I'm happy enough for it to be manual)

**Comment by willthames** [comment]
I've added two commits to your branch to improve the region handling so we don't have to maintain image ids for multiple regions forever, and so that updating compute-policy works again.

File: lib/ansible/modules/cloud/amazon/ecs_service.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ecs_service.py]: I think this can just be added to the require_if on [line 439](https://github.com/ansible/ansible/pull/40736/files#diff-5ab98c93ae8d77e786cf0afa8a0e95c3L439) with `('launch_type', 'FARGATE', ['network_configuration'])`

File: lib/ansible/modules/cloud/amazon/ecs_service.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ecs_service.py]: Rather than importing LooseVersion in the module you could use `if not module.botocore_at_least('1.8.4'):` (added a few days ago in 7f8654d586be176a355a685c249eb28fd5a17d0d)

File: lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py:303]: ~Does this need to be an int? The docs https://boto3.readthedocs.io/en/latest/reference/services/ecs.html#ECS.Client.register_task_definition indicate that~
Oops, wrong parameter. Too many 'memory' options, ignore that.

File: lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py:90]: It looks like this also accepts things like `'1GB'`

Do you mind adding an example or two to the docs?

File: lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py:319]: This could also use `module.botocore_at_least()`

Comment (mjmayer) [reply lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py:319]: It looks like the changes in 7f8654d where only for AnsibleAWSModule and ecs_taskdefinition is using AnsibleModule https://github.com/ansible/ansible/blob/6ef2ffe310b725b4f445fc0d95fff9954df94739/lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py#L239

I saw this when running a test
```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: AttributeError: 'AnsibleModule' object has no attribute 'botocore_at_least'
```

Comment (willthames) [reply lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py:319]: We should move ecs_taskdefinition to AnsibleAWSModule but we can do that in a separate PR :)

**Review by willthames — APPROVED** [review]

**Comment by willthames** [comment]
Where do the first two commits come from - normally we'd just use github's squash and merge but that would obliterate @jsteinich's contributions (I'm not too worried about my trivial fixes)

**Comment by mjmayer** [comment]
@jsteinich contributions can be found in https://github.com/PerBlue/ansible/tree/ecs_service_launch_type

**Comment by jsteinich** [comment]
I apologize for not making a pull request for my changes. I meant to but was waiting for a dependent PR to be merged.

Let me know if you need anything. I actually ended up making a couple more commits outside of the branch to fix some things I found in testing: https://github.com/PerBlue/ansible/commit/417f1fcb2c0c5d673d6075a47f3f426f0427fc58
and
https://github.com/PerBlue/ansible/commit/9b32ffdebada75f0fb55b2be91e23ac7a8e4c966

**Comment by willthames** [comment]
I'm happy for this to be merged - the remaining question is how best to do that while retaining attribution. I've got a version of your branch that squashes the whole lot into three commits to reflect the various authorships but I don't want to just force push that to your branch without asking.

**Comment by jsteinich** [comment]
@willthames If you are asking me, feel free to push a change.

I forget the exact error I got, but I needed to pass the `execution_role_arn` parameter to successfully create a task definition. It's possible that Amazon changed something since then, or maybe it has to do with the state of the AWS Account (I know some roles are automatically created under certain circumstances).

**Comment by mjmayer** [comment]
@willthames Go ahead and do the force push to my branch.

**Comment by willthames** [comment]
This is passing local tests - I'll merge once shippable passes.

**Comment by willthames** [comment]
Merged, thanks @mjmayer and @jsteinich

**Pull request merged**
